### PR TITLE
Factors out shared logic for unzipping and undeciding

### LINKED ISSUE #1: Unzipping and undeciding works for any functor
I was reading through some of the code and spotted this thing:

```haskell
-- | Distribute a 'Can' value over a product.
--
distributeCan :: Can (a,b) c -> (Can a c, Can b c)
distributeCan = \case
    Non -> (Non, Non)
    One (a,b) -> (One a, One b)
    Eno c -> (Eno c, Eno c)
    Two (a,b) c -> (Two a c, Two b c)

-- | Codistribute a coproduct over a 'Can' value.
--
codistributeCan :: Either (Can a c) (Can b c) -> Can (Either a b) c
codistributeCan = \case
    Left ac -> case ac of
      Non -> Non
      One a -> One (Left a)
      Eno c -> Eno c
      Two a c -> Two (Left a) c
    Right bc -> case bc of
      Non -> Non
      One b -> One (Right b)
      Eno c -> Eno c
      Two b c -> Two (Right b) c
```

I think we can reuse some of the other work in the module by writing:

```haskell
distributeCan :: Can (a,b) c -> (Can a c, Can b c)
distributeCan fabc = (first fst fabc, first snd fabc)

codistributeCan :: Either (Can a c) (Can b c) -> Can (Either a b) c
codistributeCan (Left fac) = first Left fac
codistributeCan (Right fbc) = first Right fbc
```

or more generally by just writing:

```haskell
unzipFirst :: Bifunctor f => f (a, b) c -> (f a c, f b c)
unzipFirst fabc = (first fst fabc, first snd fabc)

undecideFirst :: Bifunctor f => Either (f a c) (f b c) -> f (Either a b) c
undecideFirst (Left fac) = first Left fac
undecideFirst (Right fbc) = first Right fbc
```

### PR DESCRIPTION
**Pull request by masaeedu**
Fixes #1

### CODE REVIEW

**Comment by masaeedu** [comment]
Whoops, sorry. Let me just remove the nix related commit.

File: smash-core/src/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+) where
+
+import Data.Bifunctor
+
+unzipFirst :: Bifunctor f => f (a, b) c -> (f a c, f b c)
+unzipFirst fabc = (first fst fabc, first snd fabc)
+
+undecideFirst :: Bifunctor f => Either (f a c) (f b c) -> f (Either a b) c
+undecideFirst (Left fac) = first Left fac
+undecideFirst (Right fbc) = first Right fbc
```
Comment (emilypi) [inline smash-core/src/Internal.hs:13]: Good style 👍

**Comment by masaeedu** [comment]
@emilypi Ok, done.

**Comment by emilypi** [comment]
@masaeedu Thanks! Merging on green 😄

**Comment by emilypi** [comment]
Congrats @masaeedu!

**Pull request merged**